### ISSUE TITLE
Escape characters on generated Adapters

### ISSUE THREAD

**Issue by manthesepeasrock** [square/moshi#711]
There is a problem creating adapters for classes with keys that have an escape character. For example, we have this model

```kotlin
@JsonClass(generateAdapter = true)
data class EventRequest(
        @Json(name = "uri") val uri: String? = null,
        @Json(name = "event-timestamp") val timestamp: Instant,
        @Json(name = "\$type") val type: String
)
```
the key for the type should be $type, but escaping this, the adapter generated ignores the escape an since $ in kotlin strings is reserved we have Unresolved reference: type exception

Generated Class
```kotlin
// Code generated by moshi-kotlin-codegen. Do not edit.
class EventRequestJsonAdapter(moshi: Moshi) : JsonAdapter<EventRequest>() {
    private val options: JsonReader.Options =
            JsonReader.Options.of("uri", "event-timestamp", "$type")

    private val nullableStringAdapter: JsonAdapter<String?> =
            moshi.adapter(String::class.java).nullSafe()

    private val instantAdapter: JsonAdapter<Instant> = moshi.adapter(Instant::class.java).nonNull()

    private val stringAdapter: JsonAdapter<String> = moshi.adapter(String::class.java).nonNull()

    override fun toString(): String = "GeneratedJsonAdapter(EventRequest)"

    override fun fromJson(reader: JsonReader): EventRequest {
        var uri: String? = null
        var uriSet: Boolean = false
        var timestamp: Instant? = null
        var type: String? = null
        reader.beginObject()
        while (reader.hasNext()) {
            when (reader.selectName(options)) {
                0 ->  {
                    uri = nullableStringAdapter.fromJson(reader)
                    uriSet = true
                }
                1 -> timestamp = instantAdapter.fromJson(reader) ?: throw JsonDataException("Non-null value 'timestamp' was null at ${reader.path}")
                2 -> type = stringAdapter.fromJson(reader) ?: throw JsonDataException("Non-null value 'type' was null at ${reader.path}")
                -1 -> {
                    // Unknown name, skip it.
                    reader.skipName()
                    reader.skipValue()
                }
            }
        }
        reader.endObject()
        var result = EventRequest(
                timestamp = timestamp ?: throw JsonDataException("Required property 'timestamp' missing at ${reader.path}"),
                type = type ?: throw JsonDataException("Required property 'type' missing at ${reader.path}"))
        result = result.copy(
                uri = if (uriSet) uri else result.uri)
        return result
    }

    override fun toJson(writer: JsonWriter, value: EventRequest?) {
        if (value == null) {
            throw NullPointerException("value was null! Wrap in .nullSafe() to write nullable values.")
        }
        writer.beginObject()
        writer.name("uri")
        nullableStringAdapter.toJson(writer, value.uri)
        writer.name("event-timestamp")
        instantAdapter.toJson(writer, value.timestamp)
        writer.name("$type")
        stringAdapter.toJson(writer, value.type)
        writer.endObject()
    }
}
```

**Comment by NightlyNexus** [comment]
What version of Moshi?

**Comment by NightlyNexus** [comment]
Yeah. I thought we also did a workaround in Moshi 1.7 until the next KotlinPoet release?

**Comment by manthesepeasrock** [comment]
Sorry, version 1.6.0...

Will try with version 1.7 and see if the issue still stands

**Comment by manthesepeasrock** [comment]
Fixed on 1.7
